contracts/Staking.sol

```diff
@@ -64,36 +70,62 @@ contract Staking {
         }
 
         // update the next epoch pool size
-        Pool storage pNextEpoch = poolSize[tokenAddress][nextEpoch];
+        Pool storage pNextEpoch = poolSize[tokenAddress][currentEpoch + 1];
         pNextEpoch.size = token.balanceOf(address(this));
         pNextEpoch.set = true;
 
         Checkpoint[] storage checkpoints = balanceCheckpoints[msg.sender][tokenAddress];
 
+        uint256 balanceBefore = getEpochUserBalance(msg.sender, tokenAddress, currentEpoch);
+
         // if there's no checkpoint yet, it means the user didn't have any activity
         // we want to store checkpoints both for the current epoch and next epoch because
         // if a user does a withdraw, the current epoch can also be modified and
         // we don't want to insert another checkpoint in the middle of the array as that could be expensive
         if (checkpoints.length == 0) {
-            checkpoints.push(Checkpoint(currentEpoch, 0));
-            checkpoints.push(Checkpoint(nextEpoch, balances[msg.sender][tokenAddress]));
+            checkpoints.push(Checkpoint(currentEpoch, currentMultiplier, 0, amount));
+
+            // next epoch => multiplier is 1, epoch deposits is 0
+            checkpoints.push(Checkpoint(currentEpoch + 1, BASE_MULTIPLIER, amount, 0));
         } else {
             uint256 last = checkpoints.length - 1;
 
             // the last action happened in an older epoch (e.g. a deposit in epoch 3, current epoch is >=5)
             if (checkpoints[last].epochId < currentEpoch) {
-                checkpoints.push(Checkpoint(currentEpoch, checkpoints[last].balance));
-                checkpoints.push(Checkpoint(nextEpoch, balances[msg.sender][tokenAddress]));
+                checkpoints.push(Checkpoint(currentEpoch, currentMultiplier, getCheckpointEffectiveBalance(checkpoints[last]), amount));
+                checkpoints.push(Checkpoint(currentEpoch + 1, BASE_MULTIPLIER, balances[msg.sender][tokenAddress], 0));
             }
             // the last action happened in the previous epoch
```

Comment:
@kwix can you please explain this to me? It mentiones `// the last action happened in the previous epoch` but the condition is `(checkpoints[last].epochId == currentEpoch)`.

Reply:
Nvm, we can change the meaning to start of epoch balance, works fine.

Reply:
nvm, got it 👍